kytos/core/interface.py

```diff
@@ -173,19 +177,295 @@ def status_reason(self):
+            self.tag_ranges[tag_type] = tag_ranges
+
+    def remove_tag_ranges(self, tag_type: str = '1'):
+        """Remove restriction set on tag_ranges"""
+        self.tag_ranges[tag_type] = None
+
+    def get_next_available_tag(self) -> Optional[int]:
+        """Get the next available tag from the interface.
+
+        Return the next available tag if exists and remove from the
+        available tags.
+        If no tag is available return False.
+        """
+        try:
+            with self._tag_lock:
+                first_tag = self.available_tags[-1][1]
+                self.available_tags[-1][1] -= 1
+                if self.available_tags[-1][0] > self.available_tags[-1][1]:
+                    self.available_tags.pop()
+                return first_tag
+        except IndexError:
+            return False
+
+    @staticmethod
+    def find_index_remove(
+        available_tags: list[list[int]],
+        tag_range: list[int]
+    ) -> Optional[int]:
+        """Find the index of tags in available_tags to be removed"""
+        low = 0
+        high = len(available_tags)
+        while low < high:
+            mid = (low+high)//2
+            if available_tags[mid][0] <= tag_range[0]:
+                if available_tags[mid][1] >= tag_range[1]:
+                    return mid
+                low = mid + 1
+            elif available_tags[mid][0] > tag_range[0]:
+                high = mid
+        return None
+
+    def remove_tags(self, tags: list[int], tag_type: str) -> bool:
+        """Remove tags by resize available_tags
+        Returns False if nothing was remove, True otherwise"""
+        available = self.available_tags[tag_type]
+        index = self.find_index_remove(available, tags)
+        if index is None:
+            return False
+        # Resizing
+        if tags[0] == available[index][0]:
+            if tags[1] == available[index][1]:
+                available[tag_type].pop(index)
+            else:
+                available[index: index+1] = [[tags[1]+1, available[index][1]]]
+        elif tags[1] == available[index][1]:
+            available[index: index+1] = [[available[index][0], tags[0]-1]]
+        else:
+            available[index: index+1] = [
+                [available[index][0], tags[0]-1],
+                [tags[1]+1, available[index][1]]
+            ]
+        return True
+
+    def use_tags(self, tags: list[int], tag_type: str = '1') -> bool:
+        """Remove a specific tag from available_tags if it is there.
+
+        Return False in case the tags is already removed.
+        """
+        with self._tag_lock:
+            return self.remove_tags(tags, tag_type)
+
+    @staticmethod
+    def find_index_add(
+        available_tags: list[list[int]],
+        tag_range: list[int]
+    ) -> Optional[int]:
+        """Find the index of tags in available_tags to be added"""
+        if available_tags[0][0] > tag_range[1]:
+            return 0
+        low = 1
```

Comment:
I wonder if this should at 0. 

Also, looks like you want a bisect_left, instead of rolling out this custom implemention, can we use `bisect.bisect_left` instead? Same questions to all other rolled out bin search implementations

Reply:
`bisect` in [python 3.9](https://docs.python.org/3.9/library/bisect.html) only works with `list[int]`. [Python 3.10](https://docs.python.org/3.10/library/bisect.html) starts supporting `list[list[int]]` with the addition of `key=None` argument.

Reply:
@Alopalao, a `list` object since prior versions implement the dunder `__lt__`, so it should work with `bisect.bisect_*` on py3.9 too. Check it out below, if you could explore it:

```
Python 3.9.15 (main, Nov 29 2022, 00:52:25) 
[Clang 14.0.6 ] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> l = []
>>> l = [[1, 2], [4, 5]]
>>> import bisect
>>> '__lt__' in dir([])
True
>>> bisect.bisect_left(l, [3, 4])
1
>>> bisect.bisect_left(l, [0, 1])
0
>>> bisect.bisect_left(l, [6, 7])
2
>>> bisect.bisect_left(l, [4, 6])
2
>>> bisect.bisect_left(l, [4, 5])
1
>>> 

```

The only thing to be aware, is to ensure where it's called if the argument can be trusted to vale two values and sanitize/validate accordingly to avoid surprises if a wrong one is passed.

Reply:
Got it, now using `bisect.bisect_left` to obtain possible index. `Interface.find_index_remove()` has been refactored.